python/ray/_raylet.pyx

```diff
@@ -1386,6 +1385,15 @@ cdef execute_streaming_generator_sync(StreamingGeneratorExecutionContext context
     except Exception as e:
         report_streaming_generator_exception(context, e, gen_index, None)
 
+    # The caller gets object values through the reports. If we finish the task
+    # before sending the report is complete, then we may fail before the report
+    # is sent to the caller. Then, the caller would never be able to ray.get
+    # the yield'ed ObjectRef. Therefore, we must wait for all in-flight object
+    # reports to complete before finishing the task.
+    with nogil:
+        return_status = context.waiter.get().WaitAllObjectsReported()
```

Comment:
@stephanie-wang can you help me understand why we need this one? 

Provided that we apply back-pressure [here](https://github.com/ray-project/ray/blob/1070c2ad22250923a2d522e288a6b1df25123ae4/src/ray/core_worker/core_worker.cc#L3158), which should block until we get a response for `ReportGeneratorItemReturns`

Reply:
Trying to understand why this is only applicable to sync path and not an async one